README
Egg::Release::Mail.
=================================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

   or

   perl Build.PL
   ./Build
   ./Build test
   ./Build install

  * Please set the following environment variables to complete the test.
  EGG_EMAIL_ADDR    ..... Address where you may transmit mail.
  EGG_SENDMAIL_PATH ..... PATH of sendmail command.
  EGG_SCAN_HOST     ..... Host name to confirm operational condition of SMTP server.
  EGG_SCAN_PORT     ..... Port number to confirm operational condition of SMTP server.


AUTHOR

Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2008 by Bee Flag, Corp. E<lt>http://egg.bomcity.com/E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension .

0.05  2008/04/20
	- fixed for Egg::View::Mail::Plugin::Signature.
	- fixed for Egg::View::Mail::Plugin::EmbAgent.

0.04  2008/04/17
	- Build.PL is abolished.
	- fixed for Egg::View::Mail::Plugin::Signature.
	- fixed for Egg::View::Mail::Plugin::EmbAgent.

0.03  2008/04/16
	- fixed for Build.PL and Makefile.PL.

0.02  2008/03/19
	- fixed for Egg::View::Mail::Plugin::EmbAgent.

0.01  Sun Feb 24 08:28:29 2008
	- original version; created by Egg::Helper::Build::Module v3.00
	   with module name Egg::Release::Mail


lib/Egg/Release/Mail.pm
package Egg::Release::Mail;
#
# Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>
#
# $Id: Mail.pm 331 2008-04-19 16:47:33Z lushe $
#
use strict;
use warnings;

our $VERSION = '0.05';

1;

__END__

=head1 NAME

Egg::Release::Mail - Package kit for Mail Sending.

=head1 DESCRIPTION

=over 4

=item * L<Egg::View::Mail>

..... View to transmit mail.

=item * L<Egg::View::Mail::Base>

..... Base class for E-mail controller.

=item * L<Egg::View::Mail::Mailer::CMD>

..... Mail is transmitted by the sendmail command.

=item * L<Egg::View::Mail::Mailer::SMTP>

..... Mail is transmitted with L<Net::SMPT>.

=item * L<Egg::View::Mail::Encode::ISO2022JP>

..... Component for Japanese mail.

=item * L<Egg::View::Mail::MIME::Entity>

..... The content of the transmission is generated.

=item * L<Egg::View::Mail::Plugin::EmbAgent>

..... Plugin that adds client information to content of transmission.

=item * L<Egg::View::Mail::Plugin::Jfold>

..... Plug-in with which each line of content of transmission is molded in arbitrary digit.

=item * L<Egg::View::Mail::Plugin::Lot>

..... Plugin that transmits mail of this content to two or more destinations.

=item * L<Egg::View::Mail::Plugin::PortCheck>

..... The operation of the mail server is confirmed before it transmits.

=item * L<Egg::View::Mail::Plugin::Signature>

..... Plugin that adds famous etc. to content of transmission.

=item * L<Egg::Helper::View::Mail>

..... Helper to generate E-mail controller.

=back

=head1 SEE ALSO

L<Egg::Release>,
L<Net::SMTP>,

=head1 AUTHOR

Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Bee Flag, Corp. E<lt>http://egg.bomcity.com/E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Egg/View/Mail/Plugin/EmbAgent.pm
package Egg::View::Mail::Plugin::EmbAgent;
#
# Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>
#
# $Id: EmbAgent.pm 330 2008-04-19 16:42:55Z lushe $
#
use strict;
use warnings;

our $VERSION = '0.05';

sub __get_mailbody {
	my($self, $data)= @_;
	return $self->next::method($data)
	    if ($data->{no_embagent} or ! $data->{body});
	my $req   = $self->e->request;
	my $ipaddr= $req->address;
	my $regex = $data->{no_embagent_ip_regex}
	         || qr{^(?:192\.168\.|127\.0\.0\.1)};
	return $self->next::method($data) if $ipaddr=~m{$regex};
	my $body= $data->{body}= $self->__init_mailbody($data);
	$$body.= <<END_AGENT;


----------------------------------------------------------------------
END_AGENT
		if ($data->{embagent_remote_host}) {
			$$body.= <<END_AGENT;
REMOTE_HOST : @{[ $req->remote_host ]}
END_AGENT
		}
		$$body.= <<END_AGENT;
REMOTE_ADDR : @{[ $req->address ]}
USER_AGENT  : @{[ $req->agent ]}
----------------------------------------------------------------------
END_AGENT
	$body;
}

1;

__END__

=head1 NAME

Egg::View::Mail::Plugin::EmbAgent - Client information is put on the content of mail.

=head1 SYNOPSIS

  package MyApp::View::Mail::MyComp;
  use base qw/ Egg::View::Mail::Base /;
  
  ...........
  .....
  
  __PACKAGE__->setup_plugin('EmbAgent');

=head1 DESCRIPTION

Information on the access to the content of mail is put. It is MAIL plugin.

When 'EmbAgent' is passed to 'setup_plugin' method, it is built in.

There is a thing that how of information to stick changes when using it with 
other components that use '__get_mailbody' and adjust the built-in order, please.

  __PACKAGE__->setup_plugin(qw/
     Signature
     EmbAgent
     /);

The following items come to be evaluated by 'send' method of L<Egg::View::Mail::Base>
 when building it in. 

=head3 no_embagent

When an effective value is set, putting client information is canceled.

   $mail->send(
     body => .....,
     no_embagent => 1,
     );

=head3 embagent_remote_host

Information on REMOTE_HOST also sticks when an effective value is set.

   $mail->send(
     body => .....,
     embagent_remote_host => 1,
     );

Setting it in the configuration is good to always put up REMOTE_HOST.

=head1 SEE ALSO

L<Egg::Release>,
L<Egg::View::Mail>,
L<Egg::View::Mail::Base>,
L<Egg::Request>,

=head1 AUTHOR

Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 Bee Flag, Corp. E<lt>L<http://egg.bomcity.com/>E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Egg/View/Mail/Plugin/Signature.pm
package Egg::View::Mail::Plugin::Signature;
#
# Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>
#
# $Id: Signature.pm 330 2008-04-19 16:42:55Z lushe $
#
use strict;
use warnings;

our $VERSION = '0.05';

sub __get_mailbody {
	my($self, $data)= @_;
	return $self->next::method($data) unless $data->{body};
	my $body= $data->{body}= $self->__init_mailbody($data);
	$$body = $data->{body_header}. $$body if $data->{body_header};
	$$body.= "\n$data->{body_footer}"     if $data->{body_footer};
	$$body.= "\n$data->{signature}"       if $data->{signature};
	$body;
}

1;

__END__

=head1 NAME

Egg::View::Mail::Plugin::Signature - Famous etc. are added to the content of the transmission of mail. 

=head1 SYNOPSIS

  package MyApp::View::Mail::MyComp;
  use base qw/ Egg::View::Mail::Base /;
  
  ...........
  .....
  
  __PACKAGE__->setup_plugin('Signature');

=head1 DESCRIPTION

It is MAIL plugin to add famous etc. to the content when mail is transmitted.

When 'Signature' is passed to 'setup_plugin' method, it is built in.

There is a thing that processing the same as the aim is not done by the competition
 with other plugins. Please adjust the built-in order. 

  __PACKAGE__->setup_plugin(qw/
    Signature
    EmbAgent
    /);

When the following items are set by the argument and the configuration of 'send'
method, the content is added to the content of mail. 

=head3 body_header

Text added to uppermost part of content of mail.

  $mail->send( ...... , body_header => $header_text );

=head3 body_footer

Text added under content of mail.

  $mail->send( ...... , body_footer => $footer_text );

=head3 signature

Text added the under content of mail.

  $mail->send( ...... , signature => $signature_text );

=head1 SEE ALSO

L<Egg::Release>,
L<Egg::View::Mail>,
L<Egg::View::Mail::Base>,

=head1 AUTHOR

Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 Bee Flag, Corp. E<lt>L<http://egg.bomcity.com/>E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/Egg/View/Mail/Base.pm
package Egg::View::Mail::Base;
#
# Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>
#
# $Id: Base.pm 330 2008-04-19 16:42:55Z lushe $
#
use strict;
use warnings;
use Carp qw/ croak /;
use base qw/ Egg::Base Egg::Component /;

our $VERSION = '0.03';

sub setup_plugin {
	my $class= shift;
	$class->isa_register(1, "Egg::View::Mail::Plugin::$_")
	        for (ref($_[0]) eq 'ARRAY' ? @{$_[0]}: @_);
	$class->isa_terminator(__PACKAGE__);
	$class;
}
sub setup_mailer {
	my $class = shift;
	my $m_name= shift || croak q{I want mailer name.};
	my @comps;
	if (@_) { push @comps, $_ for (ref($_[0]) eq 'ARRAY' ? @{$_[0]}: @_) }
	$class->isa_register
	        (1, $_, "Egg::View::Mail::$_") for (@comps, "Mailer::$m_name");
	$class->isa_terminator(__PACKAGE__);
	$class;
}
sub setup_template {
	my $class    = shift;
	my $view_name= shift || croak q{I want view name.};
	my $template = shift || "";
	$class->mk_classdata('default_template');
	$class->default_template($template);
	no strict 'refs';  ## no critic.
	no warnings 'redefine';
	*{"${class}::___view"}=
	   sub { $_[0]->{___view} ||= $_[0]->e->view($view_name) };
	$class;
}
sub send {
	my $self= shift;
	$self->mail_send( $self->create_mail_data(@_) );
}
sub create_mail_data {
	my $self= shift;
	my $data= { %{$self->config},
	  %{ $_[0] ? ($_[1] ? {@_}: $_[0]) : croak q{I want mail data.} },
	  };
	$data->{body}= $self->create_mail_body($data);
	$data;
}
sub create_mail_body {
	my($self, $data)= @_;
	\<<END_BODY;
@{[ $self->create_mail_header($data) ]}
@{[ ${$self->__get_mailbody($data)} ]}
END_BODY
}
sub create_mail_header {
	my($self, $data)= @_;
	<<END_HEADER;
Content-Type: text/plain
To: $data->{to}
From: $data->{from}
Subject: $data->{subject}
X-Mailer: $data->{x_mailer}
END_HEADER
}
sub __init_mailbody {
	my($self, $data)= @_;
	return ($data->{template} or ! $data->{body}) ? do {
		$self->can('___view')
		     || croak q{I want setup of 'setup_template' method.};
		my $param= { %{$self->config}, %$data };
		my $tmpl = $data->{template} || $self->default_template
		     || croak q{I want 'template'.};
		my $view= $self->___view;
		@{$view->params}{keys %$param}= values %$param;
		$view->render($tmpl);
	  }: do {
		  ref($data->{body}) eq 'ARRAY'   ? \join "\n\n", @{$data->{body}}
		: ref($data->{body}) eq 'SCALAR'  ? $data->{body} : \$data->{body};
	  };
}
*__get_mailbody= \&__init_mailbody;

1;

__END__

=head1 NAME

Egg::View::Mail::Base - Base class for MAIL controller.

=head1 SYNOPSIS

  package MyApp::View::Mail::MyController;
  use base qw/ Egg::View::Mail::Base /;
  
  __PACKAGE__->config( ...... );
  
  __PACKAGE__->setup_mailer( SMTP => qw/ MIME::Entity / );

=head1 DESCRIPTION

It is a base class to succeed to from the MAIL controller who generates it in 
the helper script.

see L<Egg::Helper::View::Mail>.

=head1 METHODS

In addition, L<Egg::Base> and L<Egg::Component> have been succeeded to.

=head2 setup_plugin ([PLUGIN_LIST])

It is made to use by building the plug-in into the MAIL controller.
The name that omits the part of 'Egg::View::Mail::Plugin' is passed to PLUGIN_LIST
by the list.

  __PACKAGE__->setup_plugin(qw/
     PortCheck
     Signature
     /);

There is a thing that the competition of the method is generated by the built-in
order, it doesn't operate normally, and the result of the expectation is not
 obtained. Please adjust the built-in order to solve it.

=head2 setup_mailer ([MAILER_NAME] => [COMP_LIST])

The component that does processing that actually transmits mail is built in.

L<Egg::View::Mail::Mailer::CMD> and L<Egg::View::Mail::Mailer::SMTP> are included
 in this package.

The name that omits the part of 'Egg::View::Mail::Mailer' to MAILER_NAME is passed.

  __PACKAGE__->setup_mailer('SMTP');

In addition, other components can be specified for COMP_LIST if necessary.
Please pass the name that omits the part of 'Egg::View::Mail' by the list.

  __PACKAGE__->setup_mailer( CMD => qw/
     MIME::Entity
     Encode::ISO2022JP
     / );

There is a thing that the competition of the method is generated by the 
component as well as 'setup_plugin'.

=head2 setup_template ( [VIEW_LABEL] => [DEFAULT_TEMPLATE] )

When the content of mail is generated with a template, the template used by the
 template engine and default is set.

   __PACKAGE__->setup_template( Mason => 'mail/text.tt' );

The argument and the configuration passed to 'send' method are set in the 
parameter of the template engine. It is possible to access it by '$p' from 
among the template engine.

  To Address   : <% $p->{to} %>
  From Address : <% $p->{from} %>

=head2 send ([MAIL_DATA_HASH])

It prepares, and transmit mail processing is passed to 'mail_send' method of 
Mailer system component.

MAIL_DATA_HASH overwrites the configuration.

Therefore, the item that can be specified becomes it as well as the 
configuration.

'body' or 'template' is always necessary. 
However, this is not necessary if the template of default is set by 'setup_template'
either.

  $mail->send( body => <<END_BODY );
  Will the movie go to see ?
  END_BODY
  
     Or
  
  $mail->send( template => 'mail.tt' );

The SCALAR reference and ARRAY can be passed to 'body'.

  # Changing line enters between each element when passing it with ARRAY.
  $mail->send( body => [qw/ Will the movie go to see ? /] );
  
  # If you pass it by the SCALAR reference ...
  $mail->send( body => \"Will the movie go to see ?\n" );

=head2 create_mail_data (MAIL_DATA_HASH)

The data to do Mail Sending is made.

After merging MAIL_DATA_HASH with the configuration and making the content of 
the transmission, this method returns the data.

=head2 create_mail_body (MAIL_DATA_HASH)

The content of the transmission including the mail header is returned.

The processing of this method is not practicable. Please build in 
L<Egg::View::Mail::MIME::Entity>.

=head2 create_mail_header (MAIL_DATA_HASH)

A basic mail header is returned.

=head2 SEE ALSO

L<Egg::Release>,
L<Egg::Base>,
L<Egg::Component>,
L<Egg::View::Mail>,
L<Egg::View::Mail::MIME::Entity>,
L<Egg::Helper::View::Mail>,

=head2 AUTHOR

Masatoshi Mizuno E<lt>lusheE<64>cpan.orgE<gt>

=head2 COPYRIGHT AND LICENSE

Copyright (C) 2008 Bee Flag, Corp. E<lt>L<http://egg.bomcity.com/>E<gt>, All Rights Reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=cut

